src/tool/subcommands/api_cmd/api_run_tests.rs

```diff
@@ -0,0 +1,650 @@
+use crate::rpc::{self, RpcMethod, prelude::*};
+use crate::shim::{address::Address, message::Message};
+
+use anyhow::Context;
+use cbor4ii::core::Value;
+use cid::Cid;
+use futures::{SinkExt, StreamExt};
+use serde_json::json;
+use tokio::time::Duration;
+use tokio_tungstenite::{connect_async, tungstenite::Message as WsMessage};
+
+use std::io::{self, Write};
+use std::pin::Pin;
+use std::sync::Arc;
+
+type TestRunner = Arc<
+    dyn Fn(Arc<rpc::Client>) -> Pin<Box<dyn Future<Output = anyhow::Result<()>> + Send>>
+        + Send
+        + Sync,
+>;
+
+#[derive(Clone)]
+pub struct TestTransaction {
+    pub to: Address,
+    pub from: Address,
+    pub payload: Vec<u8>,
+    pub topic: EthHash,
+}
+
+#[derive(Clone)]
+pub struct RpcTestScenario {
+    pub run: TestRunner,
+    pub name: Option<&'static str>,
+    pub should_fail_with: Option<&'static str>,
+    pub used_methods: Vec<&'static str>,
+    pub ignore: Option<&'static str>,
+}
+
+impl RpcTestScenario {
+    /// Create a basic scenario from a simple async closure.
+    pub fn basic<F, Fut>(run_fn: F) -> Self
+    where
+        F: Fn(Arc<rpc::Client>) -> Fut + Send + Sync + 'static,
+        Fut: Future<Output = anyhow::Result<()>> + Send + 'static,
+    {
+        let run = Arc::new(move |client: Arc<rpc::Client>| {
+            Box::pin(run_fn(client)) as Pin<Box<dyn Future<Output = anyhow::Result<()>> + Send>>
+        });
+        Self {
+            run,
+            name: Default::default(),
+            should_fail_with: Default::default(),
+            used_methods: Default::default(),
+            ignore: None,
+        }
+    }
+
+    fn name(mut self, name: &'static str) -> Self {
+        self.name = Some(name);
+        self
+    }
+
+    pub fn should_fail_with(mut self, msg: &'static str) -> Self {
+        self.should_fail_with = Some(msg);
+        self
+    }
+
+    fn using<const ARITY: usize, M>(mut self) -> Self
+    where
+        M: RpcMethod<ARITY>,
+    {
+        self.used_methods.push(M::NAME);
+        if let Some(alias) = M::NAME_ALIAS {
+            self.used_methods.push(alias);
+        }
+        self
+    }
+
+    fn ignore(mut self, msg: &'static str) -> Self {
+        self.ignore = Some(msg);
+        self
+    }
+}
+
+pub(super) async fn run_tests(
+    tests: impl IntoIterator<Item = RpcTestScenario> + Clone,
+    client: impl Into<Arc<rpc::Client>>,
+    filter: String,
+) -> anyhow::Result<()> {
+    let client: Arc<rpc::Client> = client.into();
+    let mut passed = 0;
+    let mut failed = 0;
+    let mut ignored = 0;
+    let mut filtered = 0;
+
+    println!("running {} tests", tests.clone().into_iter().count());
+
+    for (i, test) in tests.into_iter().enumerate() {
+        if !filter.is_empty() && !test.used_methods.iter().any(|m| m.starts_with(&filter)) {
+            filtered += 1;
+            continue;
+        }
+        if test.ignore.is_some() {
+            ignored += 1;
+            continue;
+        }
```

Comment:
I'd print this as well, just to confirm to the developer that the methods were filtered/ignored.